pkg/apis/resourcedefinition/helper.go

```diff
@@ -0,0 +1,528 @@
+			// reuse and go ahead.
+			if lsr == nil {
+				lsr = rsr
+				continue
+			}
+
+			// Arrange the schema ref.
+			alignSchemaRef(lsr, rsr)
+		}
+
+		if lsr != nil && len(lsr.Value.Properties) != 0 {
+			ret[k] = lsr
+		}
+	}
+
+	return ret
+}
+
+// alignSchemaRef aligns the schema,
+// which always move in the direction of shrinking at first, if not found, then expanding.
+func alignSchemaRef(lsr, rsr *openapi3.SchemaRef) {
+	if lsr == nil || lsr.Value == nil ||
+		rsr == nil || rsr.Value == nil {
+		return
+	}
+
+	lv, rv := lsr.Value, rsr.Value
+
+	switch lv.Type {
+	case openapi3.TypeObject:
+		if lv.Properties != nil {
+			reqs := sets.NewString(lv.Required...)
+
+			for k := range lv.Properties {
+				// Delete the key if the type is not the same.
+				if !isSchemaRefTypeEqual(lv.Properties[k], rv.Properties[k]) {
+					delete(lv.Properties, k)
+					reqs.Delete(k)
+
+					continue
+				}
+
+				alignSchemaRef(lv.Properties[k], rv.Properties[k])
+			}
+
+			// Reset required keys.
+			if reqs.Len() == 0 {
+				lsr.Value.Required = nil
+			} else {
+				lsr.Value.Required = reqs.List()
+			}
+		} else {
+			alignSchemaRef(lv.AdditionalProperties.Schema, rv.AdditionalProperties.Schema)
+		}
+	case openapi3.TypeArray:
+		alignSchemaRef(lv.Items, rv.Items)
+	}
+
+	if lv.Format == "" && rv.Format != "" {
+		lv.Format = rv.Format
+	}
+
+	if !lv.WriteOnly && rv.WriteOnly {
+		lv.WriteOnly = rv.WriteOnly
+	}
+
+	// Find the most complete sentence.
+	// If not found, it falls back to the recommended description.
+	if mef := "description"; !isSchemaRefMutuallyExclusive(lsr, mef) {
+		switch {
+		case lv.Description == "" && rv.Description != "":
+			lv.Description = rv.Description
+		case lv.Description != "" && rv.Description != "":
+			var (
+				d      []rune
+				period int
+			)
+			lvd, rvd := []rune(lv.Description), []rune(rv.Description)
+			s := int(math.Min(float64(len(lvd)), float64(len(rvd))))
+
+			for i := 0; i < s; i++ {
+				if lvd[i] != rvd[i] {
+					break
+				}
+
+				d = append(d, lvd[i])
+
+				if lvd[i] == '.' {
+					period = i
+				}
+			}
+
+			if period != 0 {
+				lv.Description = string(d)
+				if period+1 < len(d) {
+					lv.Description = string(d[:period+1])
+				}
+			} else {
+				var rt string
+				if t := strings.ToLower(lv.Title); t != strs.Pluralize(t) {
+					rt = "Specify the " + t + "."
+				} else {
+					rt = "Specify " + t + "."
+				}
+				lv.Description = rt
+
+				setSchemaRefMutuallyExclusive(lsr, mef)
+			}
+		}
+	}
+
+	// Confirm the default value.
+	// If not the same, it falls back to no default value.
+	if mef := "default"; !isSchemaRefMutuallyExclusive(lsr, mef) {
+		if (lv.Default != nil && rv.Default == nil) ||
+			(lv.Default != nil && rv.Default != nil && !reflect.DeepEqual(lv.Default, rv.Default)) {
+			lv.Default = nil
+
+			setSchemaRefMutuallyExclusive(lsr, mef)
+		}
+	}
+
+	// Get the optimal range.
+	// If found exclusion case, it falls back to blank range.
```

Comment:
The blank range seems invalid, can we remove the property in this case?

Reply:
updated